statistics_calculations.py

```diff
@@ -1,142 +1,142 @@
-    counter = counter.astype(int)
-
-    return counter
-
-def plot_percentage(counter):
-    """
-    generates a horizontal bar plot of each category, and the percentage of each user in it.
-    counter is a df that contains the user columns + total column, and indexes of the categories
-    """
-
-    # reverse hebrew indexes
-    counter.index = counter.index.where(~counter.index.str.contains(HEBREW_LETTERS), counter.index.str[::-1])
-
-    # reverse hebrew columns
-    counter.columns = counter.columns.where(~counter.columns.str.contains(HEBREW_LETTERS), counter.columns.str[::-1])
-
-    # change counters to percentages
-    for col in counter.columns[:-1]:
-            counter[col] = counter[col].mask(counter["total"]!=0, counter[col] / counter["total"] * 100)
-
-    fig, ax = plt.subplots()
-    fig.set_figheight(len(counter) * 0.6)
-
-    # generates the first bar (starts with 0)
-    ax.barh(counter.index, counter.iloc[:,0], label=counter.columns[0])
-    for user in range(1, len(counter.columns)-1):
-        ax.barh(counter.index, counter.iloc[:,user], height=0.8, left=counter.iloc[:,:user].sum(axis=1), label=counter.columns[user])
-
-    for p in ax.patches:
-        txt = str(p.get_width().round(1)) + '%'
-
-        # locations for two users (two ends)
-        if len(counter.columns)==3:
-            if p.get_x() == 0:
-                txt_x = 0
-            else:
-                txt_x = 90
-        # if more - left side of bar
-        else:
-            txt_x = p.get_x()
-        txt_y = p.get_y()+0.4
-        ax.text(txt_x, txt_y, txt)
-
-    ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))
-    plt.xlim(0,100)
-    plt.show()
-
-def plot_word(df, word):
-    """ Counts the use of a specific word by month, and plot by user.
-        The word is passed in the form of regex.
-    """
-    # Take only messages where the word was used
-    word_df = df[df["text"].str.contains(r"{}($|\s)".format(word))]
-
-    if word_df.empty():
-        print(f"No matches found for {word}")
-        return
-
-    # Flip hebrew words to rtl
-    if word[0] in HEBREW_LETTERS:
-        word = word[::-1]
-
-    # Group by month
-    word_df = word_df.groupby([pd.Grouper(freq='M', key='date'), 'author']).count()
-    fig, ax = plt.subplots(figsize=(15, 7))
-    word_df.unstack().plot(ax=ax)
-    plt.ylim(0, plt.ylim()[1])
-    plt.title(f"Use Of {word} By User")
+import numpy as np
+import pandas as pd
+from defaults import EMOJI_REGEX, HEBREW_LETTERS
+from parsing_tools import Message
+import matplotlib.pyplot as plt
+
+def count_word(df, word):
+    """"""  # TODO - add docstring and type hints
+    """ Count use of word per author. The word is passed in the form of regex."""
+    return df[df["text"].str.contains(r"{}($|\s)".format(word))].groupby(
+        "author").count()
+
+
+def count_haha(df):
+    """"""  # TODO - add docstring and type hints
+    return df[df["text"].str.contains(r"(^|\W)ח+($|\W)")].groupby("author").count()
```

Comment:
The only issue I see is that right now, we would flag something like `בט.חחחחח`, because its using `\W` and not `\S`.
On the other hand, we want it to catch bad formatting like `סוף משפט.חחחחחח` so this is either unavoidable or would require a much more complicated regex.
Not sure we should go down that rabbit hole, what do you think?

Reply:
How often do you think בט.חחח will happen? I believe the other option is way more probable.